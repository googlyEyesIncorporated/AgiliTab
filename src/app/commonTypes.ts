import { ItemListState } from "../features/itemList/types";
import { SettingsState } from "../features/settings/types";

export interface RootState {
  itemList: ItemListState;
  settings: SettingsState;
}

export type SetBooleanState = React.Dispatch<React.SetStateAction<boolean>>;

export interface startAndEnd {
  start: number;
  end: number;
}
