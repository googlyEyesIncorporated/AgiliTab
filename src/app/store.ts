import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import itemListReducer from "../features/itemList/itemListSlice";
import settingsSlice from "../features/Settings/settingsSlice";

export const store = configureStore({
  reducer: {
    itemList: itemListReducer,
    settings: settingsSlice,
  },
  devTools: true,
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
