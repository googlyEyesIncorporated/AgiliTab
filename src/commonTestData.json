{
  "epochTimes": {
    "Sep012023": 1693540800000,
    "Sep082023": 1694188799750,
    "Sep152023": 1694836799500,
    "Sep232023": 1695484799250,
    "Sep302023": 1696132799000
  }
}
