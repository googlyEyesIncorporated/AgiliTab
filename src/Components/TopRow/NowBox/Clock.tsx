import { DateTime } from "luxon";
import { useAppSelector } from "../../../app/hooks";
import { selectTimeFormat } from "../../../features/Settings/settingsSlice";

export const Clock = ({ date }: { date: string }) => {
  const timeFormat = useAppSelector(selectTimeFormat);

  return (
    <div
      className="clock"
      style={{
        fontSize: "3em",
        width: "100%",
        border: "1px solid",
        padding: "0.5rem",
      }}
    >
      {timeFormat && DateTime.fromISO(date).toFormat(timeFormat)}
    </div>
  );
};
