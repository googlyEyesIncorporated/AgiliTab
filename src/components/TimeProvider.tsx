import { useEffect, useState, createContext, PropsWithChildren } from "react";
import { DateTime } from "luxon";

import { useAppSelector } from "../app/hooks";
import { selectTimeFormat } from "../features/settings/settingsSlice";
import { callFunctionPeriodically } from "../utils/callFunctionPeriodically";

interface ITimeProvider {
  specifiedPeriod?: number;
}
export const DateContext = createContext("");

const oneSecond = 1000;

function TimeProvider({
  specifiedPeriod = oneSecond,
  ...props
}: Readonly<PropsWithChildren<ITimeProvider>>) {
  const [date, setDate] = useState(DateTime.now().toISO() ?? "");
  const timeFormat = useAppSelector(selectTimeFormat);
  const shouldUpdateEverySecond = Boolean(timeFormat.split(":")[2]);
  const delay = shouldUpdateEverySecond ? oneSecond : specifiedPeriod;
  useEffect(() => {
    const updateDate = () => {
      setDate(DateTime.now().toISO() ?? "");
    };

    updateDate();
    return callFunctionPeriodically(delay, updateDate);
  }, [delay]);

  return <DateContext.Provider value={date} {...props} />;
}

export default TimeProvider;
