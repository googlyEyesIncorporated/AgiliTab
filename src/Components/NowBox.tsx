import { DateTime } from "luxon";
import { useEffect, useState } from "react";
import { Clock } from "./currentTimeBox/Clock";

const weekDays = [
  "Unused",
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursdays",
  "Fridays",
  "Saturday",
  "Sunday",
];

export interface ClockProps {
  h24: boolean;
}

export const NowBox = ({ h24 }: ClockProps) => {
  const [date, setDate] = useState(DateTime.now());

  useEffect(() => {
    const updateDate = () => {
      setDate(DateTime.now());
    };

    updateDate();

    const interval = setInterval(() => {
      updateDate();
    }, 1000);

    return () => clearInterval(interval);
  }, []);

  return (
    <div className="now" id="nowbox">
      <Clock date={date} h24={h24} />
      <div className="date" style={{ marginTop: "1rem" }}>
        <div>{weekDays[date.weekday]}</div>
        <div>{date.toFormat("MMM dd, yyyy")}</div>
      </div>
    </div>
  );
};
