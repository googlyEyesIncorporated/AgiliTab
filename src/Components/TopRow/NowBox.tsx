import { DateTime } from "luxon";
import { useContext, useEffect } from "react";
import { useAppDispatch, useAppSelector } from "../../app/hooks";
import {
  selectDateFormat,
  updateDay,
} from "../../features/Settings/settingsSlice";
import { Clock } from "./Clock";
import { DateContext } from "../../TimeHandler";

export const NowBox = () => {
  const dispatch = useAppDispatch();
  const dateFormat = useAppSelector(selectDateFormat);
  const date = useContext(DateContext);

  const startOfDay = DateTime.fromISO(date).startOf("day").toISO();

  // Update day start/end times
  useEffect(() => {
    dispatch(updateDay(startOfDay));
  }, [startOfDay, dispatch]);

  return (
    <div id="top-row">
      <div className="now" id="nowbox">
        <Clock date={date} />
        <span className="date" style={{ marginTop: "1rem" }}>
          <span
            style={{
              fontSize: "1.5rem",
            }}
          >
            <div>
              {DateTime.fromISO(date).weekdayLong +
                " - " +
                (dateFormat && DateTime.fromISO(date).toFormat(dateFormat))}
            </div>
          </span>
        </span>
      </div>
    </div>
  );
};
