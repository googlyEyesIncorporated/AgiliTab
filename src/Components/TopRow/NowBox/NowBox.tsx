import { DateTime } from "luxon";
import { useEffect, useState } from "react";
import { useAppDispatch, useAppSelector } from "../../../app/hooks";
import {
  selectDateFormat,
  selectTimeFormat,
  updateDay,
} from "../../../features/Settings/settingsSlice";
import { Clock } from "./Clock";

export const NowBox = () => {
  const [date, setDate] = useState(DateTime.now().toISO());
  const dispatch = useAppDispatch();
  const dateFormat = useAppSelector(selectDateFormat);
  const timeFormat = useAppSelector(selectTimeFormat);
  const withSeconds = Boolean(timeFormat.split(":")[2]);
  useEffect(() => {
    const delay = withSeconds ? 1000 : 5000;
    const updateDate = () => {
      setDate(DateTime.now().toISO());
    };

    updateDate();

    const interval = setInterval(() => {
      updateDate();
    }, delay);

    return () => clearInterval(interval);
  }, [withSeconds]);

  const startOfDay = DateTime.fromISO(date).startOf("day").toISO();

  // Update day start/end times
  useEffect(() => {
    dispatch(updateDay(startOfDay));
  }, [startOfDay, dispatch]);

  return (
    <div className="now" id="nowbox">
      <Clock date={date} />
      <div className="date" style={{ marginTop: "1rem" }}>
        <div>{DateTime.fromISO(date).weekdayLong}</div>
        <div>{dateFormat && DateTime.fromISO(date).toFormat(dateFormat)}</div>
      </div>
    </div>
  );
};
