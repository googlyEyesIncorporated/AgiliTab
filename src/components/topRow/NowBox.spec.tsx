import { render, screen } from "@testing-library/react";
import { NowBox } from "./NowBox";
import TimeProvider from "../TimeProvider";
import { epochTimes } from "../../commonTestData.json";
import { Provider } from "react-redux";
import { store } from "../../app/store";

const { Sep012023 } = epochTimes;
const jestSetTime = (newDateTime: number) => {
  jest.useFakeTimers().setSystemTime(new Date(newDateTime));
};

describe("NowBox", () => {
  beforeEach(() => {
    jestSetTime(Sep012023);
  });
  it("should render the current date/time", () => {
    render(
      <Provider store={store}>
        <TimeProvider specifiedPeriod={5000}>
          <NowBox />
        </TimeProvider>
      </Provider>
    );
    const nowBox = screen.getByTestId("now-box");
    expect(nowBox).toMatchSnapshot();
  });
});
