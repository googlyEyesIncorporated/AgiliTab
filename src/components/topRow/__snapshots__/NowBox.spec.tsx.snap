// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NowBox should render the current date/time 1`] = `
<div
  class="fade-in-down-1s flex flex-wrap lg:flex-nowrap mt-4 lg:mt-12 mb-8 justify-between"
  data-testid="now-box"
>
  <div
    class="now leading-none text-[3.375rem] border border-current border-solid my-2 mx-auto p-2 w-full lg:max-w-1/2 lg:w-3/10 text-center"
  >
    <span
      class="clock"
    >
      12:00 AM
    </span>
    <span
      class="date"
    >
      <span
        class="text-2xl"
      >
        <div>
          Friday - Sep 01, 2023
        </div>
      </span>
    </span>
  </div>
</div>
`;
